File: part-05-multi-turn-conversations/python/multi_turn_demo.py
"""
Part 5: Multi-Turn Conversations Demo
"""
import asyncio
from agent_framework.azure import AzureOpenAIResponsesClient
from azure.identity import AzureCliCredential
import os


async def multi_turn_conversation():
    """Demonstrate multi-turn conversation with thread."""
    
    # Create agent
    agent = AzureOpenAIResponsesClient(
        credential=AzureCliCredential(),
        endpoint=os.getenv("AZURE_OPENAI_ENDPOINT")
    ).create_agent(
        name="AssistantBot",
        instructions="You are a helpful assistant. Remember context from the conversation."
    )
    
    # Create a thread for this conversation
    thread = agent.get_new_thread()
    
    print("=== Multi-Turn Conversation Demo ===\n")
    
    # First turn
    print("User: My name is Alice and I'm working on a Python project.")
    result1 = await agent.run("My name is Alice and I'm working on a Python project.", thread)
    print(f"Assistant: {result1.text}\n")
    
    # Second turn - agent remembers the context
    print("User: What language am I using?")
    result2 = await agent.run("What language am I using?", thread)
    print(f"Assistant: {result2.text}\n")  # Will mention Python
    
    # Third turn - agent still has context
    print("User: What's my name again?")
    result3 = await agent.run("What's my name again?", thread)
    print(f"Assistant: {result3.text}\n")  # Will say Alice
    
    print("=== Demo Complete ===")


if __name__ == "__main__":
    asyncio.run(multi_turn_conversation())

File: part-05-multi-turn-conversations/dotnet/MultiTurnDemo.cs
using Azure.Identity;
using OpenAI;
using Microsoft.Agents.AI;

namespace MAF.Part05.MultiTurn;

/// <summary>
/// Part 5: Multi-Turn Conversations in .NET
/// </summary>
public class MultiTurnDemo
{
    public static async Task Main(string[] args)
    {
        var endpoint = Environment.GetEnvironmentVariable("AZURE_OPENAI_ENDPOINT")
            ?? throw new InvalidOperationException("AZURE_OPENAI_ENDPOINT not set");

        var agent = new AzureOpenAIClient(
                new Uri(endpoint),
                new DefaultAzureCredential())
            .GetOpenAIResponseClient("gpt-4o")
            .CreateAIAgent(
                name: "AssistantBot",
                instructions: "You are a helpful assistant. Remember context from the conversation.");

        // Create a thread for multi-turn conversation
        var thread = agent.GetNewThread();

        Console.WriteLine("=== Multi-Turn Conversation Demo ===\n");

        // First turn
        Console.WriteLine("User: I'm Bob, and I need help with Azure.");
        var result1 = await agent.RunAsync("I'm Bob, and I need help with Azure.", thread);
        Console.WriteLine($"Assistant: {result1}\n");

        // Second turn - agent remembers the context
        Console.WriteLine("User: What cloud platform am I asking about?");
        var result2 = await agent.RunAsync("What cloud platform am I asking about?", thread);
        Console.WriteLine($"Assistant: {result2}\n");  // Will mention Azure

        // Third turn - agent still has context
        Console.WriteLine("User: Remind me of my name?");
        var result3 = await agent.RunAsync("Remind me of my name?", thread);
        Console.WriteLine($"Assistant: {result3}\n");  // Will say Bob

        Console.WriteLine("=== Demo Complete ===");
    }
}
